### PR TITLE
Json annotation of mutations

### PR DESCRIPTION
**Pull request by zhx828**
There is one commit in this pr actually is in master. Once the rc has the latest master, I will rebase.

The db migration code is not included yet. No so sure what version that should go. I will just leave it to backend.

### CODE REVIEW

File: persistence/persistence-mybatis/src/main/resources/org/cbioportal/persistence/mybatis/MolecularDataMapper.xml [hunk 1]
```diff
@@ -62,6 +63,36 @@
         </where>
     </select>
 
+    <!-- This routine is a copy of getGeneMolecularAlterations above.  This copy is necessary because
```
Comment (ao508) [inline persistence/persistence-mybatis/src/main/resources/org/cbioportal/persistence/mybatis/MolecularDataMapper.xml]: is there a way that you can move all of this into a `<sql> .... </sql>` element that is referenced by these queries using `include` as it is done [here](https://github.com/cBioPortal/cbioportal/pull/6457/files#diff-4020171d42dc9dc6eb6f0c4e437a48a0R110)

File: persistence/persistence-api/src/main/java/org/cbioportal/persistence/MutationRepository.java [hunk 2]
```diff
@@ -23,6 +24,11 @@ MutationMeta getMetaMutationsInMolecularProfileBySampleListId(String molecularPr
                                                            Integer pageSize, Integer pageNumber,
                                                            String sortBy, String direction);
 
+    List<Mutation> getMutationsInMultipleMolecularProfilesByAnnotation(List<String> molecularProfileIds, List<String> sampleIds,
+                                                                       List<Integer> entrezGeneIds, String projection,
+                                                                       Integer pageSize, Integer pageNumber,
+                                                                       String sortBy, String direction, List<AnnotationFilter> filters);
+
```
[the remark this answers is not in the document]
Comment (ao508) [reply persistence/persistence-api/src/main/java/org/cbioportal/persistence/MutationRepository.java]: Another question - would annotations from GN also be grouped into the JSON field or is this more for expanding the annotations available in MAFs/the portal

File: persistence/persistence-mybatis/src/main/resources/org/cbioportal/persistence/mybatis/MolecularDataMapper.xml (hunk 1 above)
Comment (zhx828) [reply persistence/persistence-mybatis/src/main/resources/org/cbioportal/persistence/mybatis/MolecularDataMapper.xml]: @ao508 this is not from this pr.  It's introduced in https://github.com/cBioPortal/cbioportal/commit/685912300513b9b92512c54c9dfbb84080792411 I will rebase once the latest master is in rc branch.

File: persistence/persistence-mybatis/src/main/java/org/cbioportal/persistence/mybatis/MolecularDataMyBatisRepository.java [hunk 3]
```diff
@@ -36,13 +36,7 @@ public String getCommaSeparatedSampleIdsOfMolecularProfile(String molecularProfi
...
                                                                      List<Integer> entrezGeneIds, String projection) {
 
-        List<GeneMolecularAlteration> toReturn = new ArrayList();
-        Iterable<GeneMolecularAlteration> gmasItr =
-            molecularDataMapper.getGeneMolecularAlterations(molecularProfileId, entrezGeneIds, projection);
-        for (GeneMolecularAlteration gma : gmasItr) {
-            toReturn.add(gma);
-        }
-        return toReturn;
+        return molecularDataMapper.getGeneMolecularAlterations(molecularProfileId, entrezGeneIds, projection);
```
Comment (sheridancbio) [inline persistence/persistence-mybatis/src/main/java/org/cbioportal/persistence/mybatis/MolecularDataMyBatisRepository.java]: could this refactoring be done in a separate PR? It seems like it would apply independently of the changes happening in the mutation model.

File: persistence/persistence-mybatis/src/main/resources/org/cbioportal/persistence/mybatis/MolecularDataMapper.xml [hunk 4]
```diff
@@ -62,6 +63,36 @@
...
+         it is backing a corresponding method in MolecularDataMapper.java which returns a Cursor as
+         opposed to a List. This method should be kept in sync with getGeneMolecularAlterations above.
+         (Attempts where made to share getGeneMolecularAlterations between methods in MolecularDataMapper.java
```
Comment (sheridancbio) [inline persistence/persistence-mybatis/src/main/resources/org/cbioportal/persistence/mybatis/MolecularDataMapper.xml]: where = were

File: persistence/persistence-mybatis/src/main/resources/org/cbioportal/persistence/mybatis/MolecularDataMapper.xml [hunk 5]
```diff
@@ -62,6 +63,36 @@
...
+            </include>
+        </if>
+        FROM genetic_alteration
+        INNER JOIN genetic_profile ON genetic_alteration.GENETIC_PROFILE_ID = genetic_profile.GENETIC_PROFILE_ID
+        INNER JOIN gene ON genetic_alteration.GENETIC_ENTITY_ID = gene.GENETIC_ENTITY_ID
+        <where>
+            genetic_profile.STABLE_ID = #{molecularProfileId}
+            <if test="entrezGeneIds != null and !entrezGeneIds.isEmpty()">
+                AND gene.ENTREZ_GENE_ID IN
+                <foreach item="item" collection="entrezGeneIds" open="(" separator="," close=")">
```
Comment (sheridancbio) [inline persistence/persistence-mybatis/src/main/resources/org/cbioportal/persistence/mybatis/MolecularDataMapper.xml]: These "IN (.., .., ..,)" clauses can be slow. Maybe this case of selecting the genes requested by a user is not too problematic, but we also might consider doing this filtering at the service layer after querying across all genes. That way the same query is cached once and used for all searches on a particular genetic profile.

File: persistence/persistence-mybatis/src/main/resources/org/cbioportal/persistence/mybatis/MutationMapper.xml [hunk 6]
```diff
@@ -151,6 +152,61 @@
         </where>
     </sql>
 
+    <sql id="whereInMultipleMolecularProfilesByAnnotation">
+        <where>
+            <if test="sampleIds != null and !sampleIds.isEmpty()">
+                mutation.SAMPLE_ID IN (
```
Comment (sheridancbio) [inline persistence/persistence-mybatis/src/main/resources/org/cbioportal/persistence/mybatis/MutationMapper.xml]: same comment as before. Maybe we want to filter out the unwanted samples at the service layer so that we can cache and reuse a single query per profile.

File: persistence/persistence-mybatis/src/main/resources/org/cbioportal/persistence/mybatis/MutationMapper.xml [hunk 7]
```diff
@@ -151,6 +152,61 @@
...
+            </if>
+            <if test="sampleIds == null">
+                genetic_profile.STABLE_ID IN
+                <foreach item="item" collection="molecularProfileIds" open="(" separator="," close=")">
+                    #{item}
+                </foreach>
+            </if>
+            <if test="filters != null and !filters.isEmpty()">
+                <foreach item="filter" collection="filters">
+                    AND JSON_EXTRACT(mutation.annotation, #{filter.path}) in
```
Comment (sheridancbio) [inline persistence/persistence-mybatis/src/main/resources/org/cbioportal/persistence/mybatis/MutationMapper.xml]: Here is the logic where the AnnotationFilter is applied ... the JSON is extracted and we are reading out (it appears) a single field ... and looking for matches to a set of supplied values

**Comment by jjgao** [comment]
@zhx828 what's the plan with this?

**Comment by zhx828** [comment]
@jjgao never got a chance to come back work on this ticket.

**Comment by jjgao** [comment]
@zhx828 @averyniceday has this already been used in the FACETS/ASCN implementation?

**Comment by zhx828** [comment]
@jjgao I believe FACETS is using a separate column in the table. @averyniceday could comment on the details.

**Comment by averyniceday** [comment]
@zhx828 @jjgao I need to look into this PR more but off the top of my head the ASCN data does not going into the annotations column in the mutation table. It is being imported into a new allele_specific_copy_number table.

**Comment by jjgao** [comment]
@averyniceday got it. Thanks!

**Comment by jjgao** [comment]
@zhx828 I am wondering how much is left to complete this?
